src/basicfuns.jl

```diff
@@ -308,10 +308,25 @@ $(SIGNATURES)
 Return `log(exp(x) + exp(y))`, avoiding intermediate overflow/undeflow, and handling
 non-finite values.
 """
-function logaddexp(x::Real, y::Real)
-    # ensure Δ = 0 if x = y = ± Inf
-    Δ = x == y ? zero(x - y) : abs(x - y)
-    max(x, y) + log1pexp(-Δ)
+logaddexp(x::Real, y::Real) = logaddexp(float.(promote(x, y))...)
+
+function logaddexp(x::T, y::T) where T <: AbstractFloat
+    if x < y
+        diff = x - y
+        maxval = y
+    elseif y <= x
+        diff = y - x
+        maxval = x
+    else
+        diff = zero(T)
+        maxval = T(NaN)
+    end
+
+    if diff < log(eps(T)) || isnan(diff)
+        return maxval
+    else
+        return maxval + log1pexp(diff)
+    end
```

Comment:
This check is not correct, even for numbers smaller than `log(eps(T))` `log1pexp` returns non-zero values (see eg. https://github.com/JuliaStats/LogExpFunctions.jl/pull/43). With that PR it should be similarly performant (and more correct) to just use

```suggestion
    return maxval + log1pexp(diff)
```

Reply:
I've merged locally with #43 but it doesn't help that much speed wise. In this settings the main bottleneck seems to be the evaluation of the threshold. 

When using `log(eps(T))`:
```
...
329     t = log(eps(T))                                                             
330     if diff < t || isnan(diff) #log(eps(T)) || isnan(diff)                      
331         return maxval                                                           
332     else                                                                        
333         return maxval + log1pexp(diff)                                          
334     end            
335 end                                                             
```
we get:
```
BenchmarkTools.Trial: 10000 samples with 1 evaluation.                                                                     
 Range (min … max):  333.208 μs … 566.667 μs  ┊ GC (min … max): 0.00% … 40.06%                                             
 Time  (median):     334.167 μs               ┊ GC (median):    0.00%                                                      
 Time  (mean ± σ):   338.132 μs ±  28.708 μs  ┊ GC (mean ± σ):  1.12% ±  5.14%
```

and when using the "correct" threshold:
```
...
328     logtwo = oftype(diff, IrrationalConstants.logtwo)                           
329     t = (exponent(nextfloat(zero(diff))) - 1) * logtwo                          
330     if diff < t || isnan(diff) #log(eps(T)) || isnan(diff)                      
331         return maxval                                                           
332     else                                                                        
333         return maxval + log1pexp(diff)                                          
334     end                                                                         
335 end
```
we get:
```
BenchmarkTools.Trial: 4710 samples with 1 evaluation.                                                                      
 Range (min … max):  1.050 ms …  1.360 ms  ┊ GC (min … max): 0.00% … 21.33%                                                
 Time  (median):     1.054 ms              ┊ GC (median):    0.00%                                                         
 Time  (mean ± σ):   1.060 ms ± 34.891 μs  ┊ GC (mean ± σ):  0.42% ±  2.58%
```

Reply:
The threshold is not evaluated in the calls but precomputed for common types such as `Float64`, `Float32`, and `Float16`. Even for other types the compiler should usually be able to precompute the thresholds and avoid computations in every call.

Reply:
Yes, that was the point to use a predefined threshold... with the downside to lose a little bit in precision. For my particular use case (graph computation in the log-semiring), the precision loss is negligible compare to the significant gain. But this is probably too specific and it is better that the implementation in this package remains of general purpose and therefore favor precision. 

I'll simplify this part to: `return maxval + log1pexp(diff)`